src/Microsoft.AspNetCore.OData/EnableQueryAttribute.cs

```diff
@@ -38,6 +41,177 @@ public partial class EnableQueryAttribute : ActionFilterAttribute
+#if NETCOREAPP3_1 || NET5_0
+                if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(ActionResult<>))
+                {
+                    returnType = returnType.GetGenericArguments().First();
+                }
+#endif
+                if (TypeHelper.IsCollection(returnType))
+                {
+                    elementType = TypeHelper.GetImplementedIEnumerableType(returnType);
+                }
+                else if(TypeHelper.IsGenericType(returnType) && returnType.GetGenericTypeDefinition() == typeof(Task<>))
+                {
+                    elementType = returnType.GetGenericArguments().First();
+                }
+                else
+                {
+                    _queryValidationRunBeforeActionExecution = false;
+                    return;
+                }
+
+                IEdmModel edmModel = this.GetModel(
+                    elementType,
+                    request,
+                    controllerActionDescriptor);
+
+                if (controllerActionDescriptor != null)
+                {
+                    queryContext = new ODataQueryContext(
+                        edmModel,
+                        elementType);
+                    _queryValidationRunBeforeActionExecution = true;
+                }
+                else
+                {
+                    _queryValidationRunBeforeActionExecution = false;
+                    return;
+                }
+            }
+
+            // Create and validate the query options.
+            _processedQueryOptions = new ODataQueryOptions(queryContext, request);
+
+            try
+            {
+                ValidateQuery(request, _processedQueryOptions);
+            }
+            catch (ODataException odataException)
+            {
+                context.Result = CreateBadRequestResult(
+                    Error.Format(SRResources.UriQueryStringInvalid, odataException.Message),
+                    odataException);
+            }
+            catch (ArgumentOutOfRangeException e)
+            {
+                context.Result = CreateBadRequestResult(
+                    Error.Format(SRResources.QueryParameterNotSupported, e.Message),
+                    e);
+            }
+            catch (NotImplementedException e)
+            {
+                context.Result = CreateBadRequestResult(
+                    Error.Format(SRResources.UriQueryStringInvalid, e.Message),
+                    e);
+            }
+            catch (NotSupportedException e)
+            {
+                context.Result = CreateBadRequestResult(
+                    Error.Format(SRResources.UriQueryStringInvalid, e.Message),
+                    e);
+            }
+            catch (InvalidOperationException e)
+            {
+                // Will also catch ODataException here because ODataException derives from InvalidOperationException.
```

Comment:
If this is the case, do you still need a separate block for `ODataException` [here](https://github.com/OData/WebApi/pull/2362/files#diff-1cea7432b6ec6bf7cc6472857ecbea4b4d6303185cc8f6cd6aa00d6405fd5985R182)?

Reply:
Removed.